src/main/java/io/cdap/plugin/gcp/bigquery/source/BigQuerySourceUtils.java

```diff
@@ -0,0 +1,208 @@
+import io.cdap.plugin.gcp.common.GCPUtils;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.io.LongWritable;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.mapreduce.Job;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.UUID;
+import javax.annotation.Nullable;
+
+/**
+ * Utility class for BigQuery source.
+ */
+public class BigQuerySourceUtils {
+
+  private static final Logger LOG = LoggerFactory.getLogger(BigQuerySourceUtils.class);
+  public static final String GCS_BUCKET_FORMAT = "gs://%s";
+  public static final String GS_PATH_FORMAT = GCS_BUCKET_FORMAT + "/%s";
+  private static final String TEMPORARY_BUCKET_FORMAT = GS_PATH_FORMAT + "/hadoop/input/%s";
+
+  @Nullable
+  public static Credentials getCredentials(BigQueryBaseConfig config) throws IOException {
+    return config.getServiceAccount() == null ?
+      null : GCPUtils.loadServiceAccountCredentials(config.getServiceAccount(), config.isServiceAccountFilePath());
+  }
+
+  /**
+   * Gets bucket from supplied configuration.
+   *
+   * If the supplied configuration doesn't specify a bucket, a bucket will get auto created and configuration modified
+   * to auto-delete this bucket on completion.
+   *
+   * @param configuration Hadoop configuration instance.
+   * @param config BigQuery configuration.
+   * @param bigQuery BigQuery client.
+   * @param credentials Google Cloud Credentials.
+   * @param runId identified for this run.
+   * @param cmekKey CMEK key to use for the auto-created bucket.
+   * @return Bucket name.
+   */
+  public static String getOrCreateBucket(Configuration configuration,
+                                         BigQueryBaseConfig config,
+                                         BigQuery bigQuery,
+                                         Credentials credentials,
+                                         String runId,
+                                         @Nullable String cmekKey) {
+    String bucket = config.getBucket();
+
+    if (bucket == null) {
+      bucket = runId;
+      // By default, this option is false, meaning the job can not delete the bucket. So enable it only when bucket name
+      // is not provided.
+      configuration.setBoolean("fs.gs.bucket.delete.enable", true);
+
+      // the dataset existence is validated before, so this cannot be null
+      Dataset dataset = bigQuery.getDataset(config.getDataset());
+      GCPUtils.createBucket(GCPUtils.getStorage(config.getProject(), credentials), bucket, dataset.getLocation(),
+                            cmekKey);
+    }
+
+    return bucket;
+  }
+
+  /**
+   * Updates {@link Configuration} with bucket details.
+   */
+  public static void configureBucket(Configuration configuration, String bucket, String runId) {
+    configuration.set("fs.default.name", String.format("gs://%s/%s/", bucket, runId));
```

Comment:
GCS_PATH_FORMAT?

Reply:
Also, do we log this value anywhere? It makes sense to log it early to ensure we have it in case of any issues (e.g. permissions)

Reply:
You're right, this should be the static variable.